### PR TITLE
Stage 091 (datasetsRucio)

### PR DESCRIPTION
**Pull request by mgolosova**
New stage to extract additional information about datasetd from Rucio

ToDo:
- [x] update ES mapping, adding new data;
- [x] implement basic functional;
- [x] make things work in the frame of pyDKB;
- [x] adjust output (one dataset -- one message);
- [x] tell Stage 019 about different index names;
- [x] tell Stage 019 about parent/child relationship;
- [x] adjust ora2es run-script to branch dataflow...
- [x] ~~OR make things Kafka-style.~~ _(removed in favor of previous item)_
- [x] merge PR #88 
- [x] rebase on new `master` after #88 
- [x] ~~update Dataflow/README~~ _(was erroneously updated in #87)_

### CODE REVIEW

File: Utils/Elasticsearch/mapping/prodsys_rucio.mapping [hunk 1]
```diff
@@ -177,6 +173,30 @@
           }
         }
       }
+    },
+    {
+        "output_dataset": {
+            "_parent": {
```
Comment (mgolosova) [inline Utils/Elasticsearch/mapping/prodsys_rucio.mapping]: Two moments:
* broken structure: excessive `{...}` around `"output_datasets"`;
* broken style: indentation changed from 2 spaces to 4.

The structure must look this way:
```
{
  "template": "prodsys_rucio",
  "settings": {
    "number_of_shards": 4
  },
  "mappings" : {
    "task" : {
      "properties" : {
        "campaign" : {
          ...
        }
      }
    },
    "output_dataset": {
      ...
    }
  }
}

**Comment by konvpalto** [comment]
Will review this one Really Soon (tm), yeah!

**Comment by anastasiakaida** [comment]
I will review it until this Saturday.

**Comment by mgolosova** [comment]
@anastasiakaida , @konvpalto ,
thank you! Looking forward for your comments :)

**Comment by anastasiakaida** [comment]
I ran flake8 to check _datasets_processing.py_ and get this list:

`datasets_processing.py:14:1: F401 're' imported but unused
datasets_processing.py:27:80: E501 line too long (85 > 79 characters)
datasets_processing.py:45:1: E302 expected 2 blank lines, found 1
datasets_processing.py:76:1: E302 expected 2 blank lines, found 1
datasets_processing.py:102:1: E302 expected 2 blank lines, found 1
datasets_processing.py:121:5: E722 do not use bare except'
datasets_processing.py:129:1: E302 expected 2 blank lines, found 1
datasets_processing.py:141:9: F841 local variable 'result' is assigned to but never used
datasets_processing.py:141:47: E203 whitespace before ':'
datasets_processing.py:148:1: E302 expected 2 blank lines, found 1
datasets_processing.py:162:1: E302 expected 2 blank lines, found 1
datasets_processing.py:184:3: E271 multiple spaces after keyword
`
I'm not sure that all of these are important cause Travis CI checks only this: E121,E122,E126,E128,E203,E225,E226,E228,E231,E251,E261,E262,E265,E266,W291,W293,E502, but there are E203 and E502, also E501 is very importants as I know in case of convenient reading (why Travis CI exclude this codestyle error? :) ).
--------
Also there is a [small mistake](https://github.com/PanDAWMS/dkb/blob/5a5c7fa2d0c217b63d728ffa761c2f6eb1bbc134/Utils/Dataflow/091_datasetsRucio/datasets_processing.py#L10) in author's name.
--------
To be continued...

**Comment by anastasiakaida** [comment]
I ran flake8 to check _datasets_processing.py_ and got this list:

`datasets_processing.py:14:1: F401 're' imported but unused`
`datasets_processing.py:27:80: E501 line too long (85 > 79 characters)`
`datasets_processing.py:45:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:76:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:102:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:121:5: E722 do not use bare except'`
`datasets_processing.py:129:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:141:9: F841 local variable 'result' is assigned to but never used`
`datasets_processing.py:141:47: E203 whitespace before ':'`
`datasets_processing.py:148:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:162:1: E302 expected 2 blank lines, found 1`
`datasets_processing.py:184:3: E271 multiple spaces after keyword`

I'm not sure that all of these are important cause Travis CI checks only this: E121,E122,E126,E128,E203,E225,E226,E228,E231,E251,E261,E262,E265,E266,W291,W293,E502, but there are E203 and E502, also E501 is very importants as I know in case of convenient reading (why Travis CI exclude this codestyle error? :) ).
_______
Also, there is a [small mistake](https://github.com/PanDAWMS/dkb/blob/5a5c7fa2d0c217b63d728ffa761c2f6eb1bbc134/Utils/Dataflow/091_datasetsRucio/datasets_processing.py#L10) in author's name.
_______
To be continued...

**Comment by mgolosova** [comment]
@anastasiakaida,
Thank you for the check, will fix all the codestyle issues and that typo you found.

Travis doesn\`t check for lots of errors, you\`re right. It was added in a proof-of-concept mode for the actual list of errors for current PR at that moment, and I never found time to make proper settings :( But I think I`m about it.

**Comment by mgolosova** [comment]
Rebased on current version of `pr-stage-016-task2es` (5b899ae) in order to avoid excessive `"Fix codestyle"` commits.

**Comment by mgolosova** [comment]
@anastasiakaida,
have just checked the "never used" variable mentioned in your comment and found that bug in `extract_scope()`. Fixed now (in 803bcba).
Also, finaly removed the unused import of `re`.

**Comment by anastasiakaida** [comment]
@mgolosova,
Great! :)
One more question about `re`. For what purpose it was expected to be used? (I mean not full library, but expected functions).

**Comment by mgolosova** [comment]
@anastasiakaida,
Initially it was added to remove `_tidXXXXXX_XX` part from dataset name ([here](https://github.com/PanDAWMS/dkb/commit/5ff4e4012c0810229adc43274867e30b1b77862a#diff-48f0188dbe93652c09ca5c270d36766eR54)), but then the functionality was removed ([here](https://github.com/PanDAWMS/dkb/commit/2a9e734b282c65bc2c2c7d68ee1fd4e729f14e36#diff-48f0188dbe93652c09ca5c270d36766eL54)).
I joined these two commits to make a shortcut, but never noticed that import left unused.
By the way, I left links to the original commits in the joined one, and now it helped me to find the answer to your question. Conclusion: leaving references to the original history is a Good Thing (tm) ;)

And yeah, one day we should add this check for unused imports and variables to Travis as well. It will be a bit annoying (just like any other checks), but definitely useful.

File: Utils/Dataflow/019_esFormat/esFormat.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env php
+<?php
+function exception_error_handler($errno, $errstr, $errfile, $errline ) {
+   throw new ErrorException($errstr, 0, $errno, $errfile, $errline);
+}
+set_error_handler("exception_error_handler");
+
+$DEFAULT_INDEX = 'prodsys';
```
Comment (maria-grigorieva) [inline Utils/Dataflow/019_esFormat/esFormat.php]: I think here must by 'prodsys_rucio', according to updated mapping file

Comment (mgolosova) [reply Utils/Dataflow/019_esFormat/esFormat.php]: `'prodsys'` was left to correspond the original script version :)
But you`re right, it must be changed here (and kept up to date later).

**Comment by mgolosova** [comment]
The original version of the PR is available for reference in branch `pr-85-stage-91-original`.
What has changed:
* rebased on `master` (merged with the latest changes from Stage 16);
* commits 803bcba and acdb66d are joined.

Please check, and I will merge it to `master`.

**Comment by maria-grigorieva** [comment]
I've just checked stages 016 and 091. Works correctly. So, it will be merged to master. 
Thank you!

**Comment by mgolosova** [comment]
@maria-grigorieva, thank you!
Merged.

**Pull request merged**